src/data/store/slices/linkSaturation.slice.ts

```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
+                    epochState.linksPerNodeMap[nodeUid].saturation = nodeSaturation;
+                    epochState.linksPerNodeMap[nodeUid].offchipSaturation = offChipSaturation;
+                }
+            });
         });
     });
 };
 
 const updatePCILinks = (state: NetworkCongestionState) => {
     const { PCIBandwidthGBs, CLKHz } = getInitialCLKValues(state);
 
-    Object.values(state.graphs).forEach((graphState) => {
-        Object.values(graphState.links).forEach((linkState: LinkState) => {
-            if (linkState.type === LinkType.PCIE) {
-                linkState.maxBandwidth = PCIBandwidthGBs / CLKHz;
-                recalculateLinkSaturation(linkState, graphState.totalOps);
-            }
+    Object.values(state.linksPerTemporalEpoch).forEach((epochState) => {
+        Object.entries(epochState.linksPerNodeMap).forEach(([nodeUid, { links, offchipLinkIds }]) => {
+            Object.values(links).forEach((linkState) => {
+                if (linkState.type === LinkType.PCIE) {
+                    linkState.maxBandwidth = PCIBandwidthGBs / CLKHz;
+                    const { bpc, linkSaturation, nodeSaturation, offChipSaturation } = recalculateLinkSaturation({
+                        link: linkState,
+                        totalOpCycles: epochState.totalOps,
+                        links,
+                        offchipLinkIds,
+                    });
+
+                    linkState.bpc = bpc;
+                    linkState.saturation = linkSaturation;
+                    epochState.linksPerNodeMap[nodeUid].saturation = nodeSaturation;
+                    epochState.linksPerNodeMap[nodeUid].offchipSaturation = offChipSaturation;
+                }
+            });
         });
     });
 };
 
-const recalculateLinkSaturation = (link: LinkState, totalOpCycles: number) => {
-    link.bpc = link.totalDataBytes / totalOpCycles;
-    link.saturation = (link.bpc / link.maxBandwidth) * 100;
+const recalculateLinkSaturation = ({
+    link,
+    totalOpCycles,
+    links,
+    offchipLinkIds,
+}: {
+    link: LinkState;
+    totalOpCycles: number;
+    links: Record<string, LinkState>;
+    offchipLinkIds: string[];
+}) => {
+    const newLinkSaturation = (link.bpc / link.maxBandwidth) * 100;
+
+    return {
+        bpc: link.totalDataBytes / totalOpCycles,
+        linkSaturation: newLinkSaturation,
+        ...Object.entries(links).reduce(
+            (updatedSaturation, [linkId, { saturation }]) => {
+                if (linkId !== link.id) {
+                    updatedSaturation.nodeSaturation = Math.max(updatedSaturation.nodeSaturation, saturation);
+
+                    if (offchipLinkIds.includes(linkId)) {
+                        updatedSaturation.offChipSaturation = Math.max(updatedSaturation.offChipSaturation, saturation);
+                    }
+                }
+
+                return updatedSaturation;
+            },
+            {
+                nodeSaturation: newLinkSaturation,
```

Comment:
node cant be saturated or congested. nodelinks on the other hand can

Reply:
I think `nodeTotalLinkSaturation` is a little bit long though... Do you have any suggestion for the name?

Reply:
linkSaturation/linkCongestion?

Reply:
But it is for all links on the node, pre calculated